### PR TITLE
Support java modules for azure-spring-data-cosmos

### PR DESCRIPTION
**Pull request by saragluna**
Support Java modules for azure-spring-data-cosmos: https://github.com/Azure/azure-sdk-for-java/issues/14378. 

module-info.java added to the core project azure-spring-data-cosmos-core and the name for this core module is `com.azure.spring.data.cosmos`.

For the two pom file projects, `azure-spring-data-2-3-cosmos` and `azure-spring-data-2-2-cosmos`,  an `Automatic-Module-Name` entry added to the `maven-jar-plugin` so these two could be automatic modules when used in Java modules. If we don't add such entry, the automatic module names for these two are like `azure.spring.data.2.2.cosmos` which is not valid and will cause runtime exceptions.

We could choose to add `module-info.java` for the `azure-spring-data-2-3-cosmos`, but the `maven-compiler-plugin` would fail when it compiles this artifact against Java8, complaining it doesn't support Java module even though we exclude the `module-info.java` in such execution. 

I also created a WIP PR to the sample project showing the changes users need to include when creating a modular application. https://github.com/Azure-Samples/azure-spring-data-cosmos-java-sql-api-getting-started/pull/2

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/module-info.java [hunk 1]
```diff
@@ -32,7 +32,7 @@
     exports com.azure.cosmos.util;
 
     // exporting some packages specifically for Jackson
-    opens com.azure.cosmos to com.fasterxml.jackson.databind;
+    opens com.azure.cosmos to com.fasterxml.jackson.databind, org.apache.commons.lang3;
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/module-info.java:35]: Why do we need this?

Comment (saragluna) [reply sdk/cosmos/azure-cosmos/src/main/java/module-info.java:35]: We use reflection to retrieve `userAgentSuffix` field of cosmosClient: https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/cosmos/azure-spring-data-cosmos-core/src/main/java/com/azure/spring/data/cosmos/CosmosFactory.java#L87

**Comment by saragluna** [comment]
@kushagraThapar @srnagar the surefire plugin complains this exception:

> Stream 'java.lang.module.ResolutionException: Module com.azure.spring.data.cosmos contains package com.azure.cosmos, module com.azure.cosmos exports package com.azure.cosmos to com.azure.spring.data.cosmos'

But the com.azure.spring.data.cosmos does not contain package `com.azure.cosmos`, did you run into this kind of issue before?

File: sdk/cosmos/azure-cosmos/src/main/java/module-info.java (hunk 1 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/module-info.java:35]: But that code piece is in spring-data-cosmos-core and here we are exporting apache commons lang 3 to azure-cosmos module, which doesn't use this dependency at all. Let's discuss this offline today.

Comment (srnagar) [reply sdk/cosmos/azure-cosmos/src/main/java/module-info.java:35]: Yes, cosmos package should have no dependency or reference to `org.apache.commons.lang3`.

File: sdk/cosmos/azure-spring-data-cosmos-core/src/main/java/module-info.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    exports com.azure.spring.data.cosmos.core;
+    exports com.azure.spring.data.cosmos.core.mapping;
+    exports com.azure.spring.data.cosmos.exception;
+    exports com.azure.spring.data.cosmos.repository;
+    exports com.azure.spring.data.cosmos.repository.config;
+    exports com.azure.spring.data.cosmos.repository.support;
+
+    opens com.azure.spring.data.cosmos.config to spring.core;
+    opens com.azure.spring.data.cosmos.repository.support to spring.beans;
+}
```
Comment (conniey) [inline sdk/cosmos/azure-spring-data-cosmos-core/src/main/java/module-info.java:32]: nit: missing trailing empty line.

File: sdk/parents/azure-client-sdk-parent/pom.xml [hunk 3]
```diff
@@ -1142,6 +1142,18 @@
                 <!-- Azure-core needs to open the test entities to Jackson for unit tests to operate -->
                 --add-opens com.azure.core/com.azure.core.implementation.entities=com.fasterxml.jackson.databind
                 --add-opens com.azure.core/com.azure.core.implementation.entities=ALL-UNNAMED
+
```
Comment (conniey) [inline sdk/parents/azure-client-sdk-parent/pom.xml:1145]: These should not be in azure-client-sdk-parent. They are inherited by every child module. You should move these to the area path's SDK.
/cc @alzimmermsft

File: sdk/cosmos/azure-spring-data-cosmos-core/src/main/java/module-info.java (hunk 2 above)
Comment (saragluna) [reply sdk/cosmos/azure-spring-data-cosmos-core/src/main/java/module-info.java]: added

File: sdk/parents/azure-client-sdk-parent/pom.xml (hunk 3 above)
Comment (saragluna) [reply sdk/parents/azure-client-sdk-parent/pom.xml:1145]: Inherited from the child's pom.

Comment (saragluna) [reply sdk/parents/azure-client-sdk-parent/pom.xml:1145]: And there're some more `--add-opens` and `--add-exports` in the parent's pom, should they be moved to the children's poms too?

**Review by kushagraThapar — APPROVED** [review]
LGTM, thanks @saragluna 
@srnagar - can you please confirm these changes ?

File: sdk/parents/azure-client-sdk-parent/pom.xml (hunk 3 above)
Comment (kushagraThapar) [reply sdk/parents/azure-client-sdk-parent/pom.xml]: @conniey - can you please provide an update on the above discussion ?
We are trying to close this for our next release - which is early next week, and would like to get this change in by Friday (code complete)

**Review by conniey — APPROVED** [review]

**Pull request merged**